File: 2021-11-09/climbing_stairs.cpp
/**
 * https://leetcode.com/problems/climbing-stairs/
 * 
 * Solution works by using the fact that the number of ways to climb N steps
 * is equal to `number of ways to climb N - 1 steps` + `number of ways to climb
 * N - 2 steps`; the two possibilities for climbing N steps are climb N - 1
 * steps and then climb 1 step or climb N - 2 steps then climb 1 step.
 */

class Solution {
public:
    int climbStairs(int n) {
        
        int prev = 0;
        int curr = 1;
        
        while (n > 0) {
            curr += prev;
            prev = curr - prev;
            n--;
        }
        
        return curr;
    }
};

File: 2021-11-09/ClimbingStairs.java
class Solution {
    public int climbStairs(int n) {
        
       int[] steps = new int[n+1];
        steps[0] = 1;
        
        for(int i=1;i<=n;i++){
            if(i==1){
                steps[1] = steps[0];
            }else{
                steps[i] = steps[i-1] + steps[i-2];
            }
        }
        
        return steps[n];
        
    }
}
